### ISSUE TITLE
Typeahead. Ability to leaf through multiple result pages

### ISSUE THREAD

**Issue by forik** [ng-bootstrap/ng-bootstrap#1119]
It would be nice to add an option for results paging in typeahead component as there are scenarios where might be hundreds of results and it is no good to show them all at once.
Maybe it is possible to add ability to provide custom template for the whole result window popup instead of just a row template.

**Comment by trumbitta** [comment]
Also having the result window popup with internal scrolling would be nice.

Example: https://material.angularjs.org/latest/demo/autocomplete#inside-dialog

**Comment by Ks89** [comment]
I'm agree with @trumbitta.
I think that the scrollable list is the most important, in particular if you have design/UX constraints in your application.

Right now, in my application I have to put a lower bound of chars to prevent very long popups. However this limit is useless, since all my data is already into my client, so it is creating problems to users that are obliged to type 3 or more chars without a real reason.

**Comment by borgotecnologia** [comment]
Yes, it's a problem, I tried to correct it with css, but does not work, please fix this problem....

**Comment by PendelinP** [comment]
One possible workaround is to set the height of the height of the dropdown to a fixed value and enable vertical scrolling (overflow). Unfortunately the up/down-arrow-key navigation doesn't work since the focus gets lost when you select an element outside the visible area: http://plnkr.co/edit/ffZi6LSRkJUhpxA2AyvN

**Comment by ymeine** [comment]
I gave a try at implementing a solution for this issue, and here I come with a proposal.

The features implies two aspects: 

- being able to limit the height of the dropdown and scroll inside
- making sure the active item is entirely visible inside the dropdown when it is being made active

# Limiting the height

For now as this thread suggests, it is already possible to set the style of the class `dropdown-menu` and customize its `max-height` and `overflow`.

However, this might not be very handy and if we want to provide fine customization as well as a feature of automatic scrolling out of the box, I think this would make more sense to specify the maximum height to the directive directly.

Therefore I suggest an optional input property `maxHeight` (a string for maximum flexibility), forcing the dropdown's `max-height` accordingly and its `overflow` to `auto`. Setting this property to a non void value would enable the automatic scrolling feature (see below).

# Automatic scrolling

This is to make sure that the active item — the one highlighted — is always fully visible _inside the dropdown_ (as much as possible, see edge case below).

_Inside the dropdown_ means that we shouldn't take care of making sure the active item is actually visible on the screen (like `scrollIntoView` would do), it's up to the user to make the dropdown properly visible to be able to see its results. What we want is to scroll the content of the dropdown itself to make the active item visible inside of it.

The edge case is where an item would be bigger than the dropdown: in this case we would align the top of the item with the top of the dropdown.

Otherwise, the rule is to align the bottom when going down, and the top when going up.

Last but not least: this makes sense to automatically scroll in case of keyboard navigation, but I wouldn't do it in case of mouse enter, this would be disturbing. As a general rule: as long as the user uses the mouse, we leave him the responsibility for scrolling manually, but as soon as they use the keyboard we take care of automatically scrolling.

# Conclusion

As usual, I leave it open for comments for a few days, and then I'll come with a PR to implement it.

Feedback is warmly welcome :)

**Comment by rurziceanu** [comment]
No info on this? 
I rely  need this functionality as setting the overflow on the dropdown it adds scroll bars, but when you navigate with arrow key the container does not fallow the current focus item

**Comment by raygerrard** [comment]
I have run into this same issue, it would be good to have the ability to have a fixed max height while maintaining focus when navigating using the keyboard

**Comment by MickL** [comment]
Setting `max-height` and `overflow` works fine, but if the active item is outside it needs to be scrolled. 

Additionally would be neat to not set max-height manually but instead it should check if the dropdown is larger than the window and automatically set max-height then.

**Comment by ymeine** [comment]
I think all the limitations of the proposed workarounds have already been said.

As you can see on this thread __there is an open PR__, which takes care of what is requested here, without the limitations.

The problem is not this issue anymore, but rather the acceptation of the PR. There are two schools of thought though: 

- those who consider the widget should not handle such a thing since it's up to the user to provide a limited set of results (as search engines with tons of results would do)
- those who consider that it should be possible to put more results than what can be visible in one shot (i.e. without scrolling)

First group would reject the PR while second group would accept it. It's up to the integrators to decide.

@MickL the additional feature of setting a `max-height` automatically is not trivial and adding complexity to the widget is also what refrains the PR from being adopted. Moreover, I don't really know how one should handle this functionally speaking. For instance, if there is very little space below the input, do we still open the dropdown below with a small `max-height` or do we open it above? Especially since the user could scroll the parent container to reveal a bigger part of the dropdown, but with a `max-height` set to early this wouldn't change anything, so we would also have to listen to parent scrolling events to adapt the size. And there might be other issues I haven't thought about...

**Comment by pkozlowski-opensource** [comment]
> The problem is not this issue anymore, but rather the acceptation of the PR. There are two schools of thought

Totally. I think that I was pretty clear before that I'm leaning towards _non_ including this feature as IMO it makes it harder for users to make sense of results presented in typeahead. I've just tested few typeaheads / autocomplete on the major eCommerce sites and _none_ of the ones I could think of is using scrolling in autocomplete. The ones I've tested were:
* https://www.google.com - no scroll, limit of 10 results
* https://www.youtube.com/ - no scroll, limit of 10 results
* https://www.amazon.com/  - no scroll, limit of 10 results
* https://www.booking.com - no scroll, limit of 5 results
* https://www.airbnb.com - no scroll, limit of 5 results
* https://en.wikipedia.org  - no scroll, limit of 10 results

I'm definitively _not_ an usability expert but those are some of the biggest and the most used web sites in the World. 

I would love to hear more from people familiar with the topic and learn about good usability patterns where actual scrolling is required. Till we've got solid data I'm for holding off implementation of this feature (given its complexity and questionable usability).

I'm painfully aware that might make typeahead unusable for some projects and this is OK. None of the existing libraries will fill 100% of all use-cases in the World. But I would prefer that we serve 80% of people in the great way rather giving mediocre solution to 90% of people.

Anyway, this is just my thinking right now, given what I know right now. Open to be convinced!

**Comment by ymeine** [comment]
Well, unfortunately this argument has some — a bit hidden — bias. This is because of the following: 

- choosing a "big" website
- looking at their main search field

By doing that, you will alway face the same use case: __searching__ among a __huge set__ of options. In that case, it is of course not possible to think about displaying all matching results in a box, even scrolling. For consistency then, they will never do it, even if the query leads to a fair number of results that could then be displayed in a scrolling box.

However, if you use the TypeAhead more like an AutoComplete rather than a search field, you will probably more often face the case where the set of results is acceptable to be displayed in a possibly scrolling box. It can also be the case for searching, as long as this results set remains reasonably short.

Now regarding the proposed solution, I don't see what is so bad about it. Dealing with positioning and more generally pure display stuff is indeed more delicate than dealing with pure data/state. However I don't think we would expect from a GUI library to focus solely on data and DOM structure. _GUI_ includes _Graphical_, and dealing with positioning/sizing will always be part of it. I agree this is harder and more error prone. It's a bit like writing algorithms vs designing APIs. However programming often involves both.

I'm not the one to decide about the tradeoff this project should make, but this at least the knowledge I can bring to this issue.

**Comment by shabbir-dhangot** [comment]
I am facing same issue. Is there any fix coming in future?

**Comment by paso1982** [comment]
I think that the autoscrolling in key navigation is a very basic feature.. I don't really undestand why there isn't yet.. no offence.

**Comment by jpityk** [comment]
I have a simple solution to the 2 schools of thought - make it an option that you can turn on/off so that both schools are happy.  In my opinion this is a must-have in order to use this component as a form field where the user should see all available options and it needs to be scrollable.  Please let me know if making this an optional feature is feasible.

**Comment by loetscher** [comment]
I faced the same problem and imho the scrolling is a key feature which should be supported by the typeahead component itself.

Nevertheless, I created a "workaround" which works for me:
https://stackblitz.com/edit/angular-utd9ii?file=app%2Ftypeahead-scrollable.ts

**Comment by ovitrif** [comment]
It's a pity how programmers can't agree to something that will benefit the user 😂, as if they are doing these libraries for themselves

**Comment by MickL** [comment]
I wrote that i agree with this issue. But after thinking about it i think it should be closed:

It is true that a typeahead shows only a few values and nothing more. You cant scroll on Google autocomplete or on any other autocomplete i know.

**Comment by jpityk** [comment]
If a typeahead should only show a few values, then it should enforce a limit.  This typeahead compoment provides the ability to not enforce a limit, which introduces the issue at hand.

**Comment by ovitrif** [comment]
@MickL You're right, but then again this is one of the best typehead/ search dropdowns in the angular world and one that thousands of devs are likely to find useful; so maybe we can add another component to ng-bootstrap specifically for search which could be scrollable.

We're missing on a lot of happy users without one :)

**Comment by MickL** [comment]
If this is the only one i would make it its own module. Typeahead is. Ot part of Bootstrap anyway.

**Comment by ovitrif** [comment]
Personally I do agree with you, but in a real world scenario we also have testers and POs to make happy 😅, and it's not always easy to convince them

**Comment by maxtacco** [comment]
Thanks to @loetscher I was able to make typeahead scroll the way I needed:
1. Add the following css:
```css
.dropdown-menu.show {
  max-height: 400px !important;
  overflow-y: auto !important;
}
```
2. Setup typeahead:
```html
<input #searchTypeahead="ngbTypeahead"
               ...
               (keydown)="typeaheadKeydown()" />
```
3. Add the following to your component:
```typescript
  @ViewChild('searchTypeahead')
  private readonly typeahead: NgbTypeahead;
  ...
  typeaheadKeydown() {
    if (!this.typeahead.isPopupOpen()) {
      return;
    }

    setTimeout(() => {
      const popup = document.getElementById(this.typeahead.popupId),
        activeElements = popup.getElementsByClassName('active');
      if (activeElements.length === 1) {
        const elem = <HTMLElement>activeElements[0];
        elem.scrollIntoView({
          behavior: 'smooth',
          block: 'center'
        });
      }
    });
  }
```

**Comment by paso1982** [comment]
@maxtacco (also @loetscher) Thanks a lot!
Finally my customer will be satisfied.

**Comment by ocampoesteban** [comment]
Here, how I could resolve it in a easy way
 ::ng-deep.dropdown-menu.show {
    max-height: 70vh !important;
    overflow-y: auto !important;
  }

Cheers guys!

**Comment by bungeoppang** [comment]
How about this? ( extended from @maxtacco comment )
Use typeHead Element as parameter rather than combining it with ViewChild
then we can handle multiple typehead element with one function.

```html
<input #searchTypeahead="ngbTypeahead"
               ...
            (keydown)="typeaheadKeydown(searchTypeahead)" />
```
```typescript
import {NgbTypeahead} from "@ng-bootstrap/ng-bootstrap";

typeaheadKeydown(typeHeadElement) {
    if (!typeHeadElement.isPopupOpen()) {
      return;
    }

    setTimeout(() => {
      const popup = document.getElementById(typeHeadElement.popupId),
        activeElements = popup.getElementsByClassName('active');
      if (activeElements.length === 1) {
        const elem = <HTMLElement>activeElements[0];
        elem.scrollIntoView({
          behavior: 'smooth',
          block: 'center'
        });
      }
    });
}
```

**Comment by alexnoise79** [comment]
my solution, hope it helps

`<input   (keyup.arrowDown)="scroll($event)"  (keyup.arrowUp)="scroll($event)" />`

```
scroll($event: any) {
    const elem = $event.currentTarget.nextElementSibling.getElementsByClassName('active')[0];
    elem.scrollIntoView({behavior: 'smooth', block: 'nearest'});
  }
```

**Comment by projectX21** [comment]
The best feeling in my opinion I get with :
`scrollIntoView({
      behavior: 'auto',
      block: 'nearest'
  })`
